src/rgw/rgw_data_sync.cc

```diff
@@ -5202,113 +5232,175 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
+      tn->log(20, SSTR("ERROR: failed to read bucket status object. error: " << retcode));
       return set_cr_error(retcode);
     }
 
     do {
-      tn->log(20, SSTR("sync status for source bucket: " << bucket_status.state));
+      tn->log(20, SSTR("sync status for source bucket: " << bucket_status.state << 
+            ". lease is: " << (bucket_lease_cr ? "taken" : "not taken") << ". stop indications is: " << bucket_stopped));
+
+      if (bucket_status.state == BucketSyncState::Init || 
+          bucket_status.state == BucketSyncState::Stopped ||
+          bucket_stopped) { 
+        // if state is Init or Stopped, we query the remote RGW for ther state
+		    yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, sync_pair.dest_bs.bucket, &info));
+        if (retcode < 0) {
+          return set_cr_error(retcode);
+        }
+        if (info.syncstopped) {
+          // remote indicates stopped state
+          tn->log(20, "remote bilog indicates that sync was stopped");
+          if (!bucket_lease_cr) {
+            bucket_lease_cr.reset(new RGWContinuousLeaseCR(env->async_rados, env->store, status_obj,
+                                                         lock_name, lock_duration, this));
+            yield spawn(bucket_lease_cr.get(), false);
+            while (!bucket_lease_cr->is_locked()) {
+              if (bucket_lease_cr->is_done()) {
+                tn->log(5, "ERROR: failed to take bucket lease");
+                set_status("lease lock failed, early abort");
+                drain_all();
+                return set_cr_error(bucket_lease_cr->get_ret_status());
+              }
+              tn->log(5, "waiting on bucket lease");
+              yield set_sleeping(true);
+            }
+          }
+          
+          // check if local state is "stopped"
+          yield call(new ReadCR(dpp, env->async_rados, env->svc->sysobj,
+                status_obj, &bucket_status, false, &objv));
+          if (retcode < 0) {
+            tn->log(20, SSTR("ERROR: failed to read status before writing 'stopped'. error: " << retcode));
+            RELEASE_LOCK(bucket_lease_cr);
+            return set_cr_error(retcode);
+          }
+          if (bucket_status.state != BucketSyncState::Stopped) {
+            // make sure that state is changed to stopped localy
+            bucket_status.state = BucketSyncState::Stopped;
+            yield call(new WriteCR(dpp, env->async_rados, env->svc->sysobj,
+                  status_obj, bucket_status, &objv, false));
+            if (retcode < 0) {
+              tn->log(20, SSTR("ERROR: failed to write 'stopped' status. error: " << retcode));
+              RELEASE_LOCK(bucket_lease_cr);
+              return set_cr_error(retcode);
+            }
+          }
+          yield {
+            const int num_shards = sync_pipe.dest_bucket_info.layout.current_index.layout.normal.num_shards;
+            call(new RemoveBucketShardStatusCollectCR(sc, sync_pair, num_shards));
```

Comment:
we need the `generation` and `num_shards` that we were syncing, i.e.
```c++
const auto gen = bucket_status.incremental_gen;
const auto num_shards = bucket_status.shards_done_with_gen.size();
```
and we'd only expect these to be valid if `bucket_status.state` was either `Full` or `Incremental`

Reply:
we get to this code only if state is `Init` or `Incremental` with the `bucket_stopped` flag.
so, in case of `Init` we don't need to remove the files (we assume they were already removed).